### PR TITLE
Basic Tolerant CS Fixer

### PR DESCRIPTION
**Pull request by dantleech**
Have re-written the fixer logic to work on a single AST pass, and the new fixers/proposers fix some of the issues the old code had.

TODO:

- [x] Member empty line fixer
- [x] Indentation fixer
- [x] Fix docblock indentation
- [x] Reimplement partial application (only applying CS fixes to updated code)
- [ ] ...

### CODE REVIEW

**Comment by elythyr** [comment]
Seems like you fixed the issue mentioned in #15 or is the test is missing ?

**Comment by dantleech** [comment]
Added the failing test -- but I did begin to wonder if there is a better way to do this with a single pass (each fixer would be responsible for providing text edits for one node at a time (`suggestFixes(Node $node): TextEdits`)

File: lib/Adapter/TolerantParser/NodeQuery.php [hunk 1]
```diff
@@ -0,0 +1,103 @@
+<?php
+
+namespace Phpactor\CodeBuilder\Adapter\TolerantParser;
+
+use Microsoft\PhpParser\Node;
+use Microsoft\PhpParser\Node\MethodDeclaration;
+use Phpactor\CodeBuilder\Util\TextUtil;
+
+class NodeQuery
```
Comment (dantleech) [inline lib/Adapter/TolerantParser/NodeQuery.php]: final

File: lib/Adapter/TolerantParser/StyleProposer.php [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+
+namespace Phpactor\CodeBuilder\Adapter\TolerantParser;
+
+use Phpactor\CodeBuilder\Domain\TextEdits;
+
+interface StyleProposer
+{
+    public function propose(NodeQuery $node): TextEdits;
+
+    public function onExit(NodeQuery $node): TextEdits;
```
Comment (dantleech) [inline lib/Adapter/TolerantParser/StyleProposer.php]: rename proposer => onEnter

File: lib/Adapter/TolerantParser/Util/NodeHelper.php [hunk 3]
```diff
@@ -50,4 +50,17 @@ public static function resolvedShortName(Node $node, $type = null): string
...
+        if (in_array(get_class($node), $types)) {
+            $nodes[] = $node;
+        }
+
+        foreach ($node->getChildNodes() as $childNode) {
+            $nodes = self::nodesOfTypes($types, $childNode, $nodes);
+        }
+
+        return $nodes;
+    }
```
Comment (dantleech) [inline lib/Adapter/TolerantParser/Util/NodeHelper.php]: remove

File: lib/Util/TextFormat.php [hunk 4]
```diff
@@ -2,25 +2,64 @@
...
+    public function indentation(int $level): string
+    {
+        if ($level < 0) {
+            throw new RuntimeException(sprintf(
+                'Indentation level must be equal to or greater than 0, got "%s"',
+                $level
+            ));
+        }
+        return str_repeat($this->indentation, $level);
     }
```
Comment (dantleech) [inline lib/Util/TextFormat.php]: can potentially remove some of this now

File: lib/Util/TextUtil.php [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+            return $pos;
+        }
+
+        return 0;
+    }
+
+    public static function hasDocblock(string $line): bool
+    {
+        return (bool)preg_match('{^\s*\*}m', $line);
+    }
```
Comment (dantleech) [inline lib/Util/TextUtil.php]: can potentially remove some/all of this, as it's featured in the NodeQuery wrapper.

File: lib/Adapter/TolerantParser/Edits.php [hunk 6]
```diff
@@ -39,11 +41,16 @@ public function replace($node, string $text)
...
-    public function apply(string $code)
+    public function apply(string $code): string
     {
         return trim(TextEdit::applyEdits($this->edits, $code));
     }
 
+    public function textEdits(): TextEdits
+    {
+        return TextEdits::fromTextEdits($this->edits);
+    }
```
Comment (dantleech) [inline lib/Adapter/TolerantParser/Edits.php]: Can this class just be refactored away?

**Comment by dantleech** [comment]
:boom: (again) hopefully this version has less issues.

**Pull request merged**